src/Features/Core/Portable/CodeLens/CodeLensFindReferenceProgress.cs

```diff
@@ -0,0 +1,189 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeLens
+{
+    /// <summary>
+    /// Tracks incremental progress of a find references search, we use this to
+    /// count the number of references up until a certain cap is reached and cancel the search
+    /// or until the search completes, if such a cap is not reached.
+    /// </summary>
+    /// <remarks>
+    /// All public methods of this type could be called from multiple threads.
+    /// </remarks>
+    internal sealed class CodeLensFindReferencesProgress : IFindReferencesProgress
+    {
+        private readonly object _gate = new object();
+
+        /// <summary>
+        /// this token is linked to an aggregate token that is passed to the Find References
+        /// operation, this is used solely to trigger the cancellation of an in progress
+        /// find references operation, when the references count hits a given cap.
+        /// </summary>
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private readonly SyntaxNode _queriedNode;
+        private readonly ISymbol _queriedSymbol;
+        private readonly ConcurrentSet<ISymbol> _foundDefinitions = new ConcurrentSet<ISymbol>();
+
+        /// <remarks>
+        /// _referencesCount is read and written from multiple threads.
+        /// so all read and write should be through a lock around _gate.
+        /// </remarks>
+        private int _referencesCount;
```

Comment:
Not sure what advantage the `lock` gives on pure reads.  The reads are atomic by virtue of it being `int` (and hence word size).